hasktorch/test/Torch/Typed/FunctionalSpec.hs

```diff
@@ -981,6 +993,21 @@ spec' device =
           hfoldrM @IO Transpose2DSpec () (hattach cpu   (hproduct allDTypes (Proxy @'[2, 3] :. HNil)))
         Device { deviceType = CUDA, deviceIndex = 0 } ->
           hfoldrM @IO Transpose2DSpec () (hattach cuda0 (hproduct allDTypes (Proxy @'[2, 3] :. HNil)))
+      it "diag" $ do
+        let
+          shapes1 = Proxy @'[0] :. Proxy @'[1] :. Proxy @'[3] :. HNil
+          shapes2 = Proxy @'[2, 3] :. HNil
+          shapes2' = Proxy @'[0, 0] :. Proxy @'[0, 1]  :. Proxy @'[1, 0] :. HNil -- only offset 0 has upper and lower defined
```

Comment:
if certain offsets are not allowed with certain shapes, then you may want to add this to your type family above, e.g. 
```hs
type family DiagShape (index :: Nat) (shape :: [Nat]) :: [Nat] where
   ...
   DiagShape 0 '[0, n] = ... ok!
   DiagShape _ '[0, n] = TypeError "boom!"
   DiagShape 0 '[n, 0] = ... ok!
   DiagShape _ '[n, 0] = TypeError "boom!"
   ...
```

Reply:
or maybe there's a more general rule that holds for all `'[m, n]` :)

Reply:
Thanks for the tip. I think this should be addressed generally by 4523321.